lib/semian/trilogy_adapter.rb

```diff
@@ -0,0 +1,117 @@
+# frozen_string_literal: true
+
+require "semian/adapter"
+require "activerecord-trilogy-adapter"
+require "active_record/connection_adapters/trilogy_adapter"
+
+module ActiveRecord
+  module ConnectionAdapters
+    class TrilogyAdapter
+      StatementInvalid.include(::Semian::AdapterError)
+
+      class SemianError < StatementInvalid
+        def initialize(semian_identifier, *args)
+          super(*args)
+          @semian_identifier = semian_identifier
+        end
+      end
+
+      ResourceBusyError = Class.new(SemianError)
+      CircuitOpenError = Class.new(SemianError)
+    end
+  end
+end
+
+module Semian
+  module TrilogyAdapter
+    include Semian::Adapter
+
+    attr_reader :raw_semian_options, :semian_identifier
+
+    def initialize(*options)
+      *, config = options
+      @raw_semian_options = config.delete(:semian)
+      @semian_identifier = begin
+        name = semian_options && semian_options[:name]
+        unless name
+          host = config[:host] || "localhost"
+          port = config[:port] || 3306
+          name = "#{host}:#{port}"
+        end
+        :"trilogy_adapter_#{name}"
+      end
+      super
+    end
+
+    def execute(sql, name = nil, async: false)
+      if query_allowlisted?(sql)
+        super(sql, name, async: async)
+      else
+        acquire_semian_resource(adapter: :trilogy_adapter, scope: :execute) do
+          super(sql, name, async: async)
+        end
+      end
+    end
+
+    def with_resource_timeout(temp_timeout)
+      if connection.nil?
+        prev_read_timeout = @config[:read_timeout] || 0 # We're assuming defaulting the timeout to 0 won't change in Trilogy
+        @config[:read_timeout] = temp_timeout # Create new client with config
+      else
+        prev_read_timeout = connection.read_timeout
+        connection.read_timeout = temp_timeout
+      end
+      yield
+    ensure
+      @config[:read_timeout] = prev_read_timeout
+      connection&.read_timeout = prev_read_timeout
+    end
```

Comment:
Okay, so unfortunately making `half_open_resource_timeout` work is still a bit hacky, even working at the adapter layer. I'm racking my brain for another way to do this, but not sure we have many options.

Basically, the Trilogy adapter forwards configs to the client, one of which is the `read_timeout`. In the context of `with_resource_timeout`, we might be connecting to the db, in which case `connection` is nil. Consequently, the only way I see being able to set a new (temp) `read_timeout` is to mutate the `@config` so that the newly instantiated client gets created with the timeout we want. If we _have_ the connection already, we can reach directly to the connection (the `Trilogy::Client`) and modify the `read_timeout`.

In the ensure block, we can reset the timeout on the connection, and I'm assuming we should also reset the config.

This is additionally complex because if _no_ `read_timeout` is set in the config, we need to use the client's default, which I _know_ is 0 but could potentially change.